Location Reassignment Phase 3 - Part 1

mkangia opened the pull request:
Implements parts of [Phase 3 of Location Reassignment](https://docs.google.com/document/d/1h6kTr9M0FumswP3-gr_84OLeUHRmn5Y10dCpSAS3Y9g/edit?pli=1#heading=h.9m8dkkqaa6ju)

##### SUMMARY
1. reassigning cases from old location to new location, for "merge" and "move" operation only.
"extract" and "split" have an extra step to assign household cases to a specific awc id first and will be taken separately
2. update case properties form old user's commcare-user case to new user's commcare-user case
The tasks done asynchronously are done via a different queue to be able to track the progress of tasks when requests are actually performed on TCL.

##### FEATURE FLAG
`LOCATION_REASSIGNMENT`

mkangia committed:
reassign cases for merge and move operation

mkangia committed:
update new user's usercase

mkangia committed:
stickler

mkangia committed:
isort

orangejenny commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -90,3 +91,26 @@ def get_child_case_ids(ids):
+    if old_user and new_user and old_user.is_commcare_user() and new_user.is_commcare_user():
+        old_user_usercase = old_user.get_usercase()
+        new_user_usercase = new_user.get_usercase()
+        # pick values that are not already present on the new user's usercase, populated already via HQ
```
Why doesn't this also update keys that are present on both the old and new usercases but have different values?

mkangia replied:
I am not fully clear on what all updates should be done but starting with the ones that are missing. For the ones already present, my understanding is that the ones already present are set according to the new user and should not be changed?

orangejenny replied:
Ah, I get it now, yeah that sounds right.

orangejenny commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -90,3 +91,26 @@ def get_child_case_ids(ids):
+        new_user_usercase = new_user.get_usercase()
+        # pick values that are not already present on the new user's usercase, populated already via HQ
+        updates = {}
+        for key, value in old_user_usercase.case_json.items():
```
Could also use `for key in old_user_usercase.keys() - new_user_usercase.keys()` instead of these 2 lines.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/27064/commits/be0d51542eef7496a5501249fdd8770553c2a94d

orangejenny reviewed (commented):
Looks good overall, had a couple comments and questions.

mkangia committed:
merge master

mkangia commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -982,8 +982,9 @@ def get_deleted_case_ids_in_domain(domain):
     @staticmethod
-    def get_case_ids_in_domain_by_owners(domain, owner_ids, closed=None):
-        return CaseAccessorSQL._get_case_ids_in_domain(domain, owner_ids=owner_ids, is_closed=closed)
+    def get_case_ids_in_domain_by_owners(domain, owner_ids, closed=None, case_type=None):
```
I see it as if we want all case ids irrespective of closed status and mention explicitly if that is not the case. But i agree how this is confusing from rest of HQ.

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -11,14 +27,66 @@ def deprecate_locations(old_locations, new_locations, operation):
     return transition.errors
+
+
+@task(queue=settings.CELERY_LOCATION_REASSIGNMENT_QUEUE)
```
absolutely.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/27064/commits/f9be6682e6670d1bd07cbc6825874c71d97c037a

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/27064/commits/df2855a501063d903874df8be66233d5ab5c493d

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -11,14 +27,66 @@ def deprecate_locations(old_locations, new_locations, operation):
+
+
+@task(queue=settings.CELERY_LOCATION_REASSIGNMENT_QUEUE)
+def reassign_cases(domain, old_location_id, new_location_id, deprecation_time):
```
its reassigning all cases after fetching the household one, may be `reassign_household_and_child_cases_for_owner`?

orangejenny replied:
sounds good

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -90,3 +91,26 @@ def get_child_case_ids(ids):
+
+@task(queue=settings.CELERY_LOCATION_REASSIGNMENT_QUEUE)
+def update_usercase(domain, old_username, new_username):
+    if "@" not in old_username:
```
i am not sure we can be sure of it, if i remove this then next thing would be to check that in the uploaded sheet all names are without `@`.

mkangia committed:
better logic and message

mkangia committed:
move tasks to tasks

mkangia commented on settings.py:
```diff
@@ -542,6 +542,7 @@
 CELERY_REMINDER_RULE_QUEUE = 'reminder_rule_queue'
 CELERY_REMINDER_CASE_UPDATE_QUEUE = 'reminder_case_update_queue'
 CELERY_REPEAT_RECORD_QUEUE = 'repeat_record_queue'
+CELERY_LOCATION_REASSIGNMENT_QUEUE = 'celery'
```
yes!

mkangia committed:
error out in case of invalid users

mkangia committed:
less loops

mkangia committed:
fix imports for utils to tasks

mkangia commented on custom/icds/location_reassignment/parser.py:
```diff
@@ -80,6 +83,14 @@ def _parse_row(self, row, location_type_code):
                 operation, old_site_code, new_site_code
             ))
             return
+        if (
```
https://github.com/dimagi/commcare-hq/pull/27064/commits/2a122d6d0b5191910ad5d3c4422e33fc0445e1f3

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -90,3 +91,26 @@ def get_child_case_ids(ids):
+        new_username = normalize_username(new_username, domain)
+    old_user = CouchUser.get_by_username(old_username)
+    new_user = CouchUser.get_by_username(new_username)
+    if old_user and new_user and old_user.is_commcare_user() and new_user.is_commcare_user():
```
https://github.com/dimagi/commcare-hq/pull/27064/commits/df6182cec1abe123224f7c16933046bd41f9ec60

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -11,14 +27,66 @@ def deprecate_locations(old_locations, new_locations, operation):
+            submit_case_blocks(case_blocks, domain, user_id=SYSTEM_USER_ID)
+
+
+def get_household_and_child_cases_by_owner(domain, household_case_id, owner_id):
```
included in https://github.com/dimagi/commcare-hq/pull/27064/commits/f9be6682e6670d1bd07cbc6825874c71d97c037a

mkangia committed:
extract granular utils for later user

mkangia committed:
dump households

mkangia committed:
Update UI

- update form to use toggle buttons for options
- update view for form changes
- generate and email households async

mkangia committed:
update dump

- add AWC Code header
- empty Name of AWC and AWC code for user to enter
- add house hold case id to be used for reassignment

mkangia committed:
add js for select toggles

mkangia committed:
add HouseholdReassignmentParser

- also updates both parsers to return only errors

mkangia committed:
extract reassign household case

to allow individual household case reassignments

mkangia committed:
process household reassignments

mkangia committed:
add tests for HouseHolds

mkangia committed:
add tests for Reassignment Processor

mkangia committed:
move households to download

mkangia committed:
rename

mkangia committed:
merge master

mkangia committed:
isort and remove unused import

mkangia committed:
rename

mkangia committed:
continue in case of error

mkangia committed:
correct filter

mkangia committed:
items already in format needed

mkangia committed:
update comment

mkangia committed:
spacing

mkangia committed:
Merge branch 'mk/location-reassignment-5' into mk/location-reassignment-8

mkangia committed:
Merge pull request #27134 from dimagi/mk/location-reassignment-8

Location Reassignment Phase 3 - Part 2

mkangia committed:
fix import
